nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -484,83 +607,149 @@ impl DataStore {
-            Some(UpdateStatus::Updated) => true,
-            Some(UpdateStatus::NotUpdatedButExists) => false,
-            None => false,
-        };
+            .map_err(|e| {
+                // Turning all these errors into `NotFound` errors is a bit
+                // unfortunate. The query will not find anything fail if the
+                // instance ID actually doesn't exist, *or* if any of the "is
+                // it valid to set migration IDs in the current state?" checks
+                // fail, which should probably be `Error::Conflict`
+                // instead...but, we can't really tell which is the case here.
+                //
+                // TODO(eliza): Perhaps these should all be mapped to `Conflict`
+                // instead? It's arguably correct to say that trying to set
+                // migration IDs for an instance that doesn't exist is sort of a
+                // "conflict", for a significantly broad definition of "conflcit"...
+                public_error_from_diesel(
+                    e,
+                    ErrorHandler::NotFoundByLookup(
+                        ResourceType::Instance,
+                        LookupType::ById(instance_id.into_untyped_uuid()),
+                    ),
+                )
+            })
+    }
 
-        let vmm_updated = match result.vmm_status {
-            Some(UpdateStatus::Updated) => true,
-            Some(UpdateStatus::NotUpdatedButExists) => false,
-            None => false,
-        };
+    /// Unsets the migration IDs set by
+    /// [`DataStore::instance_set_migration_ids`].
+    ///
+    /// This method will only unset the instance's migration IDs if they match
+    /// the provided ones.
+    pub async fn instance_unset_migration_ids(
+        &self,
+        opctx: &OpContext,
+        instance_id: InstanceUuid,
+        migration_id: Uuid,
+        target_propolis_id: PropolisUuid,
+    ) -> Result<bool, Error> {
+        use db::schema::instance::dsl;
 
-        let migration_updated = if migration.is_some() {
-            Some(match result.migration_status {
-                Some(UpdateStatus::Updated) => true,
-                Some(UpdateStatus::NotUpdatedButExists) => false,
-                None => false,
+        let instance_id = instance_id.into_untyped_uuid();
+        let target_propolis_id = target_propolis_id.into_untyped_uuid();
+        let updated = diesel::update(dsl::instance)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::id.eq(instance_id))
+            .filter(dsl::migration_id.eq(migration_id))
+            .filter(dsl::target_propolis_id.eq(target_propolis_id))
+            .set((
+                dsl::migration_id.eq(None::<Uuid>),
+                dsl::target_propolis_id.eq(None::<Uuid>),
+                // advance the generation
+                dsl::state_generation.eq(dsl::state_generation + 1),
+                dsl::time_state_updated.eq(Utc::now()),
```

Comment:
The documentation you're referencing is for the `VmmRuntimeState` struct, but this query is mutating an _instance_ record, and in particular the fields modeled by the `InstanceRuntimeState` struct. If we look at `InstanceRuntimeState` (in `nexus/db-model/src/instance.rs`), the comment on the `gen` field says:

> The generation number for the information stored in this structure,
> including the fallback state, the instance's active Propolis ID, and its
> migration IDs.

So, I believe we must increment the generation here because we are changing the migration and target Propolis IDs --- which is what sled-agent did previously when it was responsible for doing this.

Reply:
ahhh, gotcha, I definitely misread this. Thank you - on the `instance` record, I agree, this seems legit!!